### PR TITLE
CG2 Add lazy string literal loading support

### PR DESCRIPTION
**Pull request by nattress**
* Currently all string literals are loaded as method load pre-code fixups.
* Match CG1 implementation and allow the JIT to use lazy string literal resolution so cold code strings are lazy loaded.

### CODE REVIEW

**Comment by nattress** [comment]
cc @dotnet/crossgen-contrib

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 1]
```diff
@@ -2755,8 +2755,7 @@ private void getFunctionFixedEntryPoint(CORINFO_METHOD_STRUCT_* ftn, ref CORINFO
 
         private CorInfoHelpFunc getLazyStringLiteralHelper(CORINFO_MODULE_STRUCT_* handle)
         {
-            // TODO: Lazy string literal helper
-            return CorInfoHelpFunc.CORINFO_HELP_UNDEF;
+            return CorInfoHelpFunc.CORINFO_HELP_STRCNS_CURRENT_MODULE;
```
Comment (jkotas) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2758]: Could you please move this method to `CorInfoImpl.ReadyToRun.cs` ?

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs [hunk 2]
```diff
@@ -799,6 +799,9 @@ private ISymbolNode GetHelperFtnUncached(CorInfoHelpFunc ftnNum)
                 case CorInfoHelpFunc.CORINFO_HELP_JIT_REVERSE_PINVOKE_EXIT:
                     id = ReadyToRunHelper.ReversePInvokeExit;
                     break;
+                case CorInfoHelpFunc.CORINFO_HELP_STRCNS_CURRENT_MODULE:
+                    id = ReadyToRunHelper.GetString;
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:803]: I do not think it is as simple as this. There needs to be a code to add the current module argument for the helper (it is what `ZapLazyHelperThunk` does in crossgen1.

Comment (nattress) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:803]: Ah, I thought the `CORINFO_HELP_STRCNS_CURRENT_MODULE` meant we didn't need that. I'll look into adding that lazy thunk.

Comment (trylek) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:803]: I think that ImportThunk already has a Lazy variant that I think was supposed to cater for lazy strings, among others. Having said that, it hasn't been tested until now so there's a high chance of it being broken in some way.

Comment (nattress) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:803]: Thanks Tomas - I corrected things to call the lazy thunk and fixed an x64 bug. I'll kick off a CG2 run to make sure we're okay on other architectures.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs [hunk 3]
```diff
@@ -819,6 +822,11 @@ private ISymbolNode GetHelperFtnUncached(CorInfoHelpFunc ftnNum)
             return _compilation.NodeFactory.GetReadyToRunHelperCell(id);
         }
 
+        private CorInfoHelpFunc getLazyStringLiteralHelper(CORINFO_MODULE_STRUCT_* handle)
+        {
+            return CorInfoHelpFunc.CORINFO_HELP_STRCNS_CURRENT_MODULE;
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:827]: Is this right for strings from other modules? Crossgen1 version of this method has a check for current module in this method.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs (hunk 2 above)
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:803]:
```suggestion
```

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs (hunk 3 above)
Comment (nattress) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:826]: I believe that's a fragile ngen scenario. `getLazyStringLiteralHelper` gets called from the below handling code in the JIT. Returning `CORINFO_HELP_STRCNS` causes the JIT to call `embedModuleHandle` which is not supported in R2R.
https://github.com/dotnet/runtime/blob/1390941e4f1076209b434bb4d0cda034ba9febdf/src/coreclr/jit/morph.cpp#L9287-L9301

Comment (jkotas) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:826]: Does this all work well for large version bubble or composite mode cases?

**Review by trylek — APPROVED** [review]
LGTM, thank you! Could you please address JanK's concerns by running a private composite Crossgen2 lab run on the change?

**Comment by nattress** [comment]
I'm abandoning this change. I initially made it to reduce file size; however the generated code is larger than the savings from removing a method pre-code fixup. In their current form lazy string literals are only used for string literals in catch blocks so the optimization is a marginal one anyway.

**Comment by jkotas** [comment]
> lazy string literals are only used for string literals in catch blocks

Lazy string literals are used in throw blocks. They are startup time and working set optimization. We may want to measure how much startup time and working set we are giving up by not implementing them.

**Pull request closed without merging**